src/gurt/shm_alloc.c

```diff
@@ -0,0 +1,331 @@
+		goto err;
+	}
+	d_shm_head        = (struct d_shm_hdr *)shm_addr;
+	d_shm_head->magic = 0;
+	/* initialize memory allocators */
+	for (i = 0; i < N_SHM_POOL; i++) {
+		d_shm_head->tlsf[i] = tlsf_create_with_pool(
+			shm_addr + sizeof(struct d_shm_hdr) + (i * shm_pool_size), shm_pool_size);
+	}
+
+	if (shm_mutex_init(&(d_shm_head->g_lock)) != 0) {
+		DS_ERROR(errno, "shm_mutex_init() failed");
+		goto err_unmap;
+	}
+	for (i = 0; i < N_SHM_POOL; i++) {
+		if (shm_mutex_init(&(d_shm_head->mem_lock[i])) != 0) {
+			DS_ERROR(errno, "shm_mutex_init() failed");
+			goto err_unmap;
+		}
+	}
+	if (shm_mutex_init(&(d_shm_head->ht_lock)) != 0) {
+		DS_ERROR(errno, "shm_mutex_init() failed");
+		goto err_unmap;
+	}
+
+	pid_shm_creator         = pid;
+	d_shm_head->off_ht_head = INVALID_OFFSET;
+
+	atomic_store_relaxed(&(d_shm_head->ref_count), 1);
+	atomic_store_relaxed(&(d_shm_head->large_mem_count), 0);
+	d_shm_head->size           = shm_size;
+	d_shm_head->shm_pool_size  = shm_pool_size;
+	d_shm_head->magic          = DSM_MAGIC;
+	/* initialization is finished now. */
+	return 0;
+
+err_unmap:
+	d_shm_head = NULL;
+	munmap(shm_addr, shm_size);
+
+err:
+	close(shm_ht_fd);
+	return errno;
+}
+
+int
+shm_init(void)
+{
+	int      i;
+	int      shm_ht_fd;
+	int      shmopen_perm = 0600;
+	void    *shm_addr;
+	int      rc;
+	char     daos_shm_name_buf[64];
+	uint64_t shm_size;
+	uint64_t shm_pool_size;
+
+	if (page_size == 0)
+		page_size = sysconf(_SC_PAGESIZE);
+	if (pid == 0)
+		pid = getpid();
+	if (d_shm_head) {
+		while (d_shm_head->magic != DSM_MAGIC)
+			usleep(1);
+		/* shared memory already initlized in current process */
+		return 0;
+	}
+
+	rc = d_getenv_uint64_t("DAOS_SHM_SIZE", &shm_size);
+	if (rc != -DER_NONEXIST) {
+		/* set parameter from env */
+		shm_pool_size = shm_size / N_SHM_POOL;
+		if (shm_pool_size % page_size)
+			/* make shm_pool_size 4K aligned */
+			shm_pool_size += (page_size - (shm_pool_size % page_size));
+		shm_size = shm_pool_size * N_SHM_POOL + sizeof(struct d_shm_hdr);
+	} else {
+		shm_pool_size = SHM_POOL_SIZE;
+		shm_size      = SHM_SIZE_REQ;
+	}
+
+	rc = pthread_mutexattr_init(&d_shm_mutex_attr);
+	D_ASSERT(rc == 0);
+	rc = pthread_mutexattr_settype(&d_shm_mutex_attr, PTHREAD_MUTEX_NORMAL);
+	D_ASSERT(rc == 0);
+	rc = pthread_mutexattr_setpshared(&d_shm_mutex_attr, PTHREAD_PROCESS_SHARED);
+	D_ASSERT(rc == 0);
+	pthread_mutexattr_setrobust(&d_shm_mutex_attr, PTHREAD_MUTEX_ROBUST);
+	D_ASSERT(rc == 0);
+
+	/* the shared memory only accessible for individual user for now */
+	sprintf(daos_shm_name_buf, "%s_%d", daos_shm_name, getuid());
+open_rw:
+	shm_ht_fd = shm_open(daos_shm_name_buf, O_RDWR, shmopen_perm);
+	/* failed to open */
+	if (shm_ht_fd == -1) {
+		if (errno == ENOENT) {
+			rc = create_shm_region(shm_size, shm_pool_size);
+			if (rc == EEXIST)
+				goto open_rw;
+			else
+				return rc;
+		} else {
+			DS_ERROR(errno, "unexpected error shm_open()");
```

Comment:
im curious why the need for retry in this case?
also this should be a warn not an error in this case.

Reply:
I thought retry might make the code more robust.
I just went through all possible errno returned by shm_open(). Retry should be not needed. I will change DS_ERROR to DS_WARN. Thank you!

Reply:
The retry code was removed. DS_ERROR is kept to log the error of shm_open().